prog/encoding.go

```diff
@@ -1017,7 +1017,11 @@ func (p *parser) fixupAutos(prog *Prog) {
 				_ = s
 			case *PtrType:
 				a := arg.(*PointerArg)
-				a.Address = s.ma.alloc(nil, a.Res.Size())
+				var alignment uint64 = 1
+				if structType, ok := a.Res.Type().(*StructType); ok && structType.AlignAttr != 0 {
```

Comment:
I think this won't work correctly for e.g. X in:
```
X {
  y Y
}

Y {
  z int32
} [align[1024]]
```

This piece of code is also duplicated 3 times. We have function that computes type alignment in pkg/compiler:comp.typeAlign. I think we need to move it to prog package and maybe refactor it into Type method. Then we can call it nicely as `s.ma.alloc(nil, a.Res.Size(), a.Res.Type().Align())`.

Reply:
I've started in making this happen.
I've included it as a separate commit for now to keep things separated, I'll squash things when we are happy with how this looks.
Is this more or less what you had in mind?
Meanwhile I'll be working out the int64Alignment where I call it with 0 currently, as it depends on the actual target.

Reply:
Yes, this is roughly what I meant.
However, I see the problem with int64Alignment and passing 0 and passing this int64Alignment argument everywhere is just a bit ugly.
To solve all of that we somehow need to make Type self-contained and be able to figure own alignment without access to the target.
Taking into account that eventually we want ability to specify custom alignment attribute for all types:
https://github.com/google/syzkaller/issues/245#issuecomment-566998551
it probably makes sense to add TypeAlign directly to TypeCommon. Then we don't need AlignAttr/PackedAttr or structs separately, and we don't need int64Alignment, and in this case we don't actually need to move typeAlign from pkg/compiler to prog. pkg/compiler could calculate and store alignment for each type based on the type, target and any attributes.

Reply:
@dvyukov new approach:
Added TypeCommon.TypeAlign. Type.Alignment([no arg]) is to be used to call alloc.
For int types, TypeAlign is set in Gen in compiler/types.go.
For struct where AttrAlign was used, now we use TypeAlign.

But I'm not sure on setting TypeAlign in Gen right away for Types that contain other Types (Struct/Union or Array).
Currently it calls the underlying elements in Type.Alignment.
Should that part go to compiler/types.go as well to always just return TypeAlign?

Then we can instead alloc with t.TypeAlign instead of t.Alignment()